src/main/java/com/_up/megastore/services/implementations/ProductService.java

```diff
@@ -50,12 +50,21 @@ public Product findProductByIdOrThrowException(UUID productId) {
     }
 
     @Override
-    public void deleteProductById(UUID productId){
+    public void deleteProduct(UUID productId){
+        ifProductIsNotDeletedThrowException(productId);
         Product product = findProductByIdOrThrowException(productId);
         product.setDeleted(true);
         productRepository.save(product);
     }
 
+    @Override
+    public void ifProductIsNotDeletedThrowException(UUID productId){
+        Product product = findProductByIdOrThrowException(productId);
+        if (product.isDeleted()){
+            throw new IllegalStateException("Product with id " + productId + " is already deleted.");
```

Comment:
No está mal pero hay una forma más óptima de hacerlo: con Query Methods. Podrías probar implementando la lógica de lo siguiente
```suggestion
        if (productRepository.existsByProductIdAndDeletedTrue(productId)) {
            throw new IllegalStateException("Product with id " + productId + " is already deleted");
```
Dentro del IProductRepository